[logging] Introducing structured log API

andll opened the pull request:
This PR introduces structured log API for libra.

# Summary

In a nutshell, API introduces two things
1) `StructuredLogEntry` class and `send_struct_log!` macro for directly composing structured log
2) Bridge between traditional log! macro and structured logging API

# Direct API

```rust
pub struct StructuredLogEntry {
    text: Option<String>,
    pattern: Option<&'static str>,
    name: Option<&'static str>,
    module: Option<&'static str>,
    location: Option<&'static str>,
    git_rev: Option<&'static str>,
    data: HashMap<&'static str, Value>,
}

impl StructuredLogEntry {
    pub fn new_unnamed() -> Self { /* ... */ }
    pub fn new_named(name: &'static str) -> Self { /* ... */ }
    /* ..Bunch of builder style setters for chained initialization such as
        entry.data(a, b).data(x, y).. */
}

// Usage:
send_struct_log!(StructuredLogEntry::new_named("Committed")
   .data("block", &block)
   .data("autor", &author))
```

Arguments passed to `.data` will be serialized into json, and as such should implement `Serialize`.
Only static strings are allowed as field names.

`send_struct_log!` should be used to send structured log entry.
This macro populates metadata such as git_rev, location and module, and also skips evaluation of StructuredLogEntry entirely, if structured logging is disabled

# Log macro bridge

Crate owners are not required to rewrite their code right away to support new structured logging.
Importing logger crate will automatically emit structured logging on every log(debug!, info!, etc) macro invocation, unless `no_struct_log` feature was enabled on this crate.

*Note*: Only enable 'no_struct_log' on leaf binary crates, never on library, as it might disable structured logging for other crates, due to how cargo handles features at this moment.

So
```
info!("Committing {}", block);
```
Will emit(in addition to regular text log) structured log such as
```
{
  pattern: "Committing {}",
  data: {
    block: "<id>"
  },
  ...metadata...
}
```

There are few caveats to automatic structured logging generation

1) Argument values for structured logging will be serialized with `Debug` implementation(vs proper json serialization if `send_struct_log!` macro is used), regardless of what formatter is used for textual log. As a consequence, this means, that every log argument must implement `Debug`. In theory, this is extra limitation, but in practice currently in libra node crate and it's dependencies there was no single log argument that would not satisfy this criteria.

2) Field names will be automatically evaluated if expression is a single identifier, as in example above field `block` will get human readable name in json. However, if more complex expression is passed, structured log field name will be based on position of argument: "_0", "_1", etc:
```
info!("Committing {}", block.id());
//->
{
  data: {
    "_0": "<id>"
  },
  ...metadata...
}
```

3) Another way to set proper name to json fields is to use named format arguments:
```
info!("Committing {id}", id=block.id());
//->
{
  data: {
    "id": "<id>"
  },
  ...metadata...
}
```

# Structured log sink

Application must define implementation of `StructLogSink` interface in order to direct structured logs emitted by `send_struct_log` and other macros. This sink can be only initialized once, by calling `set_struct_logger` function.

Currently 3 implementations for `StructLogSink` exist:

1) `NopStructLog` ignores structured logs
2) `PrintStructLog` immediately prints structured logs to stdout
3) `FileStructLog` prints structured logs into provided file. Using this logger creates separate thread for writing files and structured logging itself is asynchronous in this case.

bors-libra commented:
:umbrella: The latest upstream changes (presumably d6a2f89527f1843b36780141679796fed6f76a2f) made this pull request unmergeable. Please resolve the merge conflicts.

andll commented:
Upd: Added `FileStructLog`

andll commented:
So my current thought is that structured logs would have to go to separate file, and we would need separate logstash config to pull them
Motivation for that is that we would have something in stdout/err anyway, like panic messages, etc, so mixing in structured logs there would make things break from time to time.

@sherry-x @davidiw  What do you guys think?

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
 pub const DEFAULT_TARGET: &str = "libra";
 const RUST_LOG: &str = "RUST_LOG";
 
+pub trait StructLogSink: Sync {
```
maybe we should separate out this file into two modules:
structured.rs
unstructured.rs
and maybe
macros.rs

and only include here our re-exports?

andll replied:
Even with comments/documentation this lib.rs is still under 1000 lines.
For comparison, rust log crate has lib.rs of 1.6k lines. Is it really necessary to split?

davidiw replied:
My personal preference is to split based upon the level of modularity and only merge multiple modules if it ends up being < 100 ~ 200 lines of code. This will effectively contain 3 different modules and 1 major feature (structured logging), so it seems relatively straightforward on how one would break it up into pieces. We are also likely going to extend upon the complexity within this crate.

With respect to the rust log crate, I did not enjoy my time with that crate. I spent far more time searching side of it (via ctrl-f) than I would like to admit. Ultimately it made me better at navigating large, single files within Rust, but definitely not a flavor I prefer.

I'll defer to you, but I would really like to see this split up.

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
+
+#[derive(Default, Serialize)]
+pub struct StructuredLogEntry {
+    #[serde(skip_serializing_if = "Option::is_none")]
```
what is the behavior in JSON if the option is None? At least in TOML it doesn't serialize the line, so these macros would be redundant if we were using TOML.

andll replied:
Default behavior puts None into json

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
+}
+
+// This is exact copy of similar function in log crate
+pub fn set_struct_logger(logger: &'static dyn StructLogSink) -> Result<(), ()> {
```
This makes sense for MVP, but I would think at some point we might want to have a common entry point to our thread that takes in a log and it then routes the structured logging to the structured logging handler and the unstructured to the log handler. Thoughts?

That way we could also re-initialize our internal logger as many times we want with different filter levels etc etc.

andll replied:
Yeah, we'll have entry point that initializes both loggers based on configuration / env variables

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
+}
+
+#[must_use = "use StructuredLogEntry::send to send structured log"]
+impl StructuredLogEntry {
```
have you considered making the fields above public? Does it make sense for this struct to be public?

andll replied:
This struct has to be public since it is initialized with macro

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
+#[cfg(not(feature = "no_struct_log"))]
+macro_rules! struct_log_enabled {
+    ($level:expr) => {
+        $crate::log::log_enabled!(target: $crate::DEFAULT_TARGET, $level)
```
This seems a bit different than how Log does it, right? Do they expect us to handle this at the actually log(record) function? Or else where?

andll replied:
For now we just check log level from main log crate + whether struct log is enabled at all
We can adjust implementation/arguments of this macro later as we have better picture how we want to control structured loggin

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
+    ($entry:expr) => {
+        if $crate::struct_logger_enabled() {
+            let mut entry = $entry;
+            entry.timestamp($crate::Utc::now().format("%F %T").to_string());
```
Why aren't we using machine readable timestamps for structured logs? Do we anticipate humans would be reading the json or in a  presenter?

andll replied:
At least one tailer that I saw (fluentbit) support any timestamp format. If reader can parse human readable format, why make it un-readable?

davidiw commented on network/Cargo.toml:
```diff
@@ -32,6 +32,7 @@ libra-config = { path = "../config", version = "0.1.0" }
 libra-crypto = { path = "../crypto/crypto", version = "0.1.0" }
 libra-types = { path = "../types", version = "0.1.0" }
 libra-logger = { path = "../common/logger", version = "0.1.0" }
+#libra-logger = { path = "../common/logger", version = "0.1.0", features=["names_required"] }
```
?

davidiw commented on common/logger/Cargo.toml:
```diff
@@ -15,3 +15,9 @@ edition = "2018"
+serde = { version = "1.0.96", features = ["derive"] }
+
+[features]
+names_required = []
```
shouldn't this be default and than disabled via no-default otherwise?

andll replied:
Again, due to how cargo handles features, this is not as easy as it seems...

andll commented on common/logger/Cargo.toml:
```diff
@@ -15,3 +15,9 @@ edition = "2018"
+
+[features]
+names_required = []
+no_struct_log = []
```
Yeah, I started with this, but unfortunately this does not seem to work properly with cargo feature propagation

andll commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
-#![forbid(unsafe_code)]
-
-use chrono::Utc;
+pub use chrono::Utc;
```
struct log macro uses it :)

andll commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,426 @@
+pub use chrono::Utc;
 use env_logger::filter;
 pub use log::{self, Level, Log, Metadata, Record};
+pub use serde_json;
```
Same, macro uses it

andll commented:
Added crate level documentation(majority of it duplicates this PR's description) and fixed few nits(network crate cargo and `pub use Writer`)

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,531 @@
+    ($($arg:tt)+) => {
+        let mut entry = $crate::StructuredLogEntry::new_unnamed();
+        $crate::format_struct_args_and_pattern!(entry, $($arg)+);
+        $crate::send_struct_log!(entry);
```
I see now, why not take the approach taken by the log crate, where they have a very thin macro that only inlines the needed code and forks off otherwise, e.g.,

```
macro_rules! struct_log {
  ($($arg::tt)+) => {
    crate::send_structured_log(module!(), file!(), line!(), $crate::parse_arguments($(arg)+));
  }
}
```

That way you only expose the function send_struct_log to the outside world?

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,531 @@
+#[macro_export]
+macro_rules! send_struct_log {
+    ($entry:expr) => {
+        if $crate::struct_logger_enabled() {
```
we could probably move this into a function and hide entry...

 I don't know the Rust philosophy, but my intuition is to eliminate as much from within macros as possible.

andll replied:
This has to be macro so we can skip evaluation when logging is disabled

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,74 +1,531 @@
+}
+
+#[macro_export]
+macro_rules! format_struct_args {
```
This seems a lot more complicated then what they did in log's structured log case:
```
(target: $target:expr, $lvl:expr, ($($arg:expr),*) { $($key:ident : $value:expr),* }) => {{
        let lvl = log::Level::Info;
        if lvl <= $crate::STATIC_MAX_LEVEL && lvl <= $crate::max_level() {
            $crate::__private_api_log(
                __log_format_args!($($arg),*),
                lvl,
                &(__log_module_path!(), __log_module_path!(), __log_file!(), __log_line!()),
                Some(&[$((__log_stringify!($key), &$value)),*])
```
I guess its because of the flexibility you were trying to obtain?

andll replied:
Logging macro piggy backs on macro from `fmt`, while we need to do formatting ourselves, thats why it is more complex

davidiw replied:
We should have a macro competition to see if someone can minimize this while still maintaining readability.

sherry-x commented:
This is awesome!! Logstash works pretty well with json format, so this should be good!
How do we make the structure log store in a separate file? Currently we just direct all the log stream from docker to a local file, and I don't think we can direct output based on log type.

andll commented:
* Split implementations into struct_log.rs and text_log.rs. Macros are kept in lib.rs, which is normal practice
* Removed all new `pub use` re-exports

davidiw commented on common/logger/src/struct_log.rs:
```diff
@@ -0,0 +1,211 @@
+        ret
+    }
+
+    pub fn text(&mut self, text: String) -> &mut Self {
```
I've seen several different ways of implementing setters in Rust, how did you land on this one?
Versus
fn set_*(mut self, ...) -> Self
fn set_*(&mut self, ...) -> &self
or some variant thereof...

andll replied:
This one seem to be most flexible.
`set(mut self)` would require to use chaining or re-assign variables constantly

davidiw commented on common/logger/src/lib.rs:
```diff
@@ -1,363 +1,319 @@
-pub const DEFAULT_TARGET: &str = "libra";
-const RUST_LOG: &str = "RUST_LOG";
+mod struct_log;
+pub use struct_log::*;
```
if possible be explicit on re-exports

andll commented on common/logger/src/struct_log.rs:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    // Use send_struct_log! macro instead of this method to populate extra meta information such as git rev and module name
+    pub fn send(self) {
```
Thats a good idea

andll commented:
Addressed feedback

@bors-libra r=davidiw

bors-libra commented:
:pushpin: Commit 66431b9 has been approved by `davidiw`

<!-- @bors-libra r=davidiw 66431b93b32d418326379574f18431600eeb1e3b -->

andll commented:
@bors-libra r=davidiw

bors-libra commented:
:pushpin: Commit 5018961 has been approved by `davidiw`

<!-- @bors-libra r=davidiw 5018961a6ab3bba766ca741b2e7e61491dcb005c -->

bors-libra commented:
:hourglass: Testing commit 5018961a6ab3bba766ca741b2e7e61491dcb005c with merge 9e658575cd3d87a7f4ecf16b416b40a38a268e43...

andll commented:
@bors-libra r=davidiw

bors-libra commented:
:pushpin: Commit c4a418d has been approved by `davidiw`

<!-- @bors-libra r=davidiw c4a418dc96337aa6885c0c717cbdc699428417b9 -->

andll commented:
@bors-libra r=davidiw

bors-libra commented:
:pushpin: Commit 5f39b73 has been approved by `davidiw`

<!-- @bors-libra r=davidiw 5f39b731cbd1eaf00fa514fbc75535a096dfae7d -->

bors-libra commented:
:hourglass: Testing commit 5f39b731cbd1eaf00fa514fbc75535a096dfae7d with merge 4c017aa74295d2a5f6bf15a0f14d71928c6297ca...

bors-libra commented:
:sunny: Test successful - [checks-actions_land_blocking_test](https://github.com/libra/libra/runs/520330713), [checks-circle_commit_workflow](https://circleci.com/workflow-run/c46a14bc-7f89-48d5-a02a-be6709f60e4b?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: davidiw
Pushing 4c017aa74295d2a5f6bf15a0f14d71928c6297ca to master...
